### PR TITLE
Easy perf improvement for GolomRiceFilters

### PR DESCRIPTION
**Pull request by NicolasDorier**
Ping @lontivero some low hanging fruits for better perfs of golomb filters.
Should relief the GC a bit.

### CODE REVIEW

File: NBitcoin/BIP158/GolombRiceFilter.cs [hunk 1]
```diff
@@ -481,7 +483,7 @@ public GolombRiceFilter Build()
 			return new GolombRiceFilter(filterData, n, _p, _m);
 		}
 
-		private static byte[] Compress(List<ulong> values, byte P)
+		private static byte[] Compress(ulong[] values, byte P)
```
Comment (knocte) [inline NBitcoin/BIP158/GolombRiceFilter.cs]: here you could use IEnumerable?

Comment (NicolasDorier) [reply NBitcoin/BIP158/GolombRiceFilter.cs]: meh, it is private method only used here, and a performance sensitive part. Not really worth it.

File: NBitcoin/BIP158/GolombRiceFilter.cs [hunk 2]
```diff
@@ -398,7 +399,8 @@ public GolombRiceFilterBuilder AddScriptPubkey(Script scriptPubkey)
 			if (scriptPubkey == null)
 				throw new ArgumentNullException(nameof(scriptPubkey));
 
-			_values.Add(scriptPubkey.ToBytes());
+			// Unsafe is OK because Script is readonly and we do not modify the arrays inside values
+			_values.Add(scriptPubkey.ToBytes(true));
```
Comment (lontivero) [inline NBitcoin/BIP158/GolombRiceFilter.cs]: Yes! this has to be good.

**Comment by lontivero** [comment]
I've review the commits and they look very good. Currently operating with compact filter in NBitcoin is quite fast, and will be faster with these changes. However, cf requires a lot of bit level operations and that's where processing 200GB bit by bit can be a slow. That's something could be improved but it is not causing any problem in Wasabi, in fact I think it will never become a problem.

The last time I measured them I got the numbers below. Take into account that in Wasabi those times has to be necessarily better because we only use p2wpkh scripts what makes the filter much smaller.

```
Filters generation time: 00:00:02.9757793
Filters matching time: 00:00:00.4871420
Filters matching false time: 00:00:01.3413390
```

However, after that there were an improvement in the `SipHash` that is called millions of times (and here @NicolasDorier made it a struct :+1: ) so, those numbers have to be even better.

File: NBitcoin/BIP158/GolombRiceFilter.cs [hunk 3]
```diff
@@ -155,7 +154,18 @@ public uint256 GetHeader(uint256 previousHeader)
...
+
+		/// <summary>
+		/// Checks if any of the provided elements is in the filter.
+		/// </summary>
+		/// <param name="data">Data elements to check in the filter.</param>
+		/// <param name="key">Key used for hashing the data elements.</param>
+		/// <returns>true if at least one of the elements is in the filter, otherwise false.</returns>
+		public bool MatchAny(byte[][] data, byte[] key)
+		{
+			return MatchAny(data, data.Length, key);
```
Comment (lontivero) [inline NBitcoin/BIP158/GolombRiceFilter.cs]: Possible NRE in `data.Length`

File: NBitcoin/BIP158/GolombRiceFilter.cs [hunk 4]
```diff
@@ -166,10 +176,32 @@ public bool Match(byte[] data, byte[] key)
...
+			{
+				return MatchAny(dataArray, dataArray.Length, key);
+			}
+			else if (data is ICollection<byte[]> dataCollection)
+			{
+				return MatchAny(dataCollection, dataCollection.Count, key);
+			}
+			else
+			{
+				return MatchAny(data, data.Count(), key);
```
Comment (lontivero) [inline NBitcoin/BIP158/GolombRiceFilter.cs]: Possible NRE.

**Comment by lontivero** [comment]
ACK. I review it again and don't see anything that can break. Just the two possible NRE that were not introduced in this refactoring it seems.

**Pull request merged**